### PR TITLE
Made linreg completely generic.

### PR DESCRIPTION
**Pull request by cseed**
Other regressions will be forthcoming.

### CODE REVIEW

File: src/main/scala/is/hail/stats/RegressionUtils.scala [hunk 1]
```diff
@@ -185,6 +233,27 @@ object RegressionUtils {
       None
   }
 
+  def parseXExpr(xExpr: String, ec: EvalContext): () => java.lang.Double = {
+    val (xt, xf0) = Parser.parseExpr(xExpr, ec)
+
+    def castToDouble[T](f: (T) => Double): () => java.lang.Double = { () =>
```
Comment (jbloom22) [inline src/main/scala/is/hail/stats/RegressionUtils.scala]: There's duplication between the logic here and the logic in `toDouble` and `getSampleAnnotation[s]` just below where toDouble is used, so consider pulling that out.

There's another duplication is logically separate from this PR: getSampleAnnotation is a special case of getSampleAnnotations and getPhenoCovCompleteSamples is a special case of getPhenosCovCompleteSamples. I left the former in when creating the latter a long time ago so as not to add extra nesting overhead. Do you think I should remove them?

File: src/main/scala/is/hail/stats/RegressionUtils.scala [hunk 2]
```diff
@@ -268,15 +337,15 @@ object RegressionUtils {
...
-    val yIS = getSampleAnnotations(vds, yExpr, ec)
-    val covIS = getSampleAnnotations(vds, covExpr, ec)
+    val yIS = getSampleAnnotations(vsm, yExpr, ec)
+    val covIS = getSampleAnnotations(vsm, covExpr, ec)
 
     val (yForCompleteSamples, covForCompleteSamples, completeSamples) =
-      (yIS, covIS, 0 until vds.nSamples)
+      (yIS, covIS, 0 until vsm.nSamples)
         .zipped
         .filter((y, c, s) => y.forall(_.isDefined) && c.forall(_.isDefined))
```
[the remark this answers is not in the document]
Comment (cseed) [reply src/main/scala/is/hail/stats/RegressionUtils.scala]: Why?  This seems fine.

File: src/main/scala/is/hail/stats/RegressionUtils.scala [hunk 3]
```diff
@@ -268,15 +337,15 @@ object RegressionUtils {
...
+      "sa" -> (1, vsm.saSignature))
 
     val ec = EvalContext(symTab)
 
-    val yIS = getSampleAnnotations(vds, yExpr, ec)
-    val covIS = getSampleAnnotations(vds, covExpr, ec)
+    val yIS = getSampleAnnotations(vsm, yExpr, ec)
+    val covIS = getSampleAnnotations(vsm, covExpr, ec)
 
     val (yForCompleteSamples, covForCompleteSamples, completeSamples) =
```
[the remark this answers is not in the document]
Comment (cseed) [reply src/main/scala/is/hail/stats/RegressionUtils.scala]: > Similarly, the function names should be getPhenoCovCompleteSampleIndex and getPhenosCovCompleteSampleIndex.

This seems outside the scope of this change.  Also, the current names seem better to me, which I read as "get the phenotypes and covariates (for) the complete samples".

File: src/main/scala/is/hail/stats/RegressionUtils.scala (hunk 1 above)
Comment (cseed) [reply src/main/scala/is/hail/stats/RegressionUtils.scala]: Probably be fine to remove them, but I might leave it for when you go back to the generalized regression stuff.

**Comment by cseed** [comment]
OK, back to you.  I'll rebase and address the rest of the stack once we're happy with this one.

**Comment by jbloom22** [comment]
Did you intend to push a new commit to address the first few comments?

**Comment by cseed** [comment]
Oops, pushed the wrong repo. Done.

**Comment by jbloom22** [comment]
why `convExpr` instead of `covExpr`?

**Comment by cseed** [comment]
Oops, typo, fixed.

**Pull request merged**